### ISSUE TITLE
Schema validation does not show actual problem

### ISSUE THREAD

**Issue by afreudenreich** [simonkowallik/as3ninja#74]
I am using the schema validation in code as follows:
`as3schema = AS3Schema()`
`as3schema.validate(declaration=my_declaration)`
The validation works and throws an AS3ValidationError as expected when something is wrong.

In my test example one of the properties had a comma at the end:
"abc.test.com.80,": {
"class": "Service_HTTP",

But the error itself does not tell me where something is wrong, as it just states following:

_Validation error: data/output/declarations/dev/dev.service.json ..._ (shows the full declaration here)
_Failed validating 'oneOf' in schema:..._
(and here follows the complete schema and the full declaration again, but no hint which property might actually be wrong).

Is this expected behaviour or am I missing something?

Thanks for any hints,
Andreas

**Comment by simonkowallik** [comment]
@afreudenreich thanks for submitting this issue. I'll look into it. I believe the python jsonschema lib has something to offer. It will not work in all cases I guess, but it looks promising for this particular case.

**Comment by afreudenreich** [comment]
@simonkowallik thanks for looking into this :)


I have encountered other examples in case you are interested or it helps with troubleshooting:

1) Faulty declaration:
`"snat": {
    "bigip": {
        "default_pool": "/Common/default_pool"
    }
}`

Results in following error:
`
...
File ".../code/bigip/as3-configuration-manager/venv/lib/python3.10/site-packages/as3ninja/schema/formatcheckers.py", line 108, in _regex_match
    if reg_ex.match(value) is None:
TypeError: expected string or bytes-like object`

The same declaration using the dry-run check of the F5 Plugin in VS Code:
`"results": [
        {
            "message": "Invalid data property: default_pool",
            "code": 422
        }`

2) Fault declaration:
`"PTR_SNAT_Pool": "general_pool2",` (the object is not used)

Results in the same huge error as given in the last post.

The same declaration using the dry-run check of the F5 Plugin in VS Code:
```
"message": "Invalid data property: PTR_SNAT_Pool",
"code": 422
```

3) Faulty declaration:
`"snat": "PTR_SNAT_Pool"`

Results in the same huge error as given in the last post

The same declaration using the dry-run check of the F5 Plugin in VS Code:
```
"errors": [
    "/abc.test.cybersecurity.ubc.ca/web.443/abc.test.cybersecurity.ubc.ca.443/snat: should be equal to one of the allowed values [\"none\",\"self\",\"auto\"]"
```
 
In summary: I am not sure if F5 have written their own schema validator, but their output is consistently very specific. I would expect that jsonschema lib would be able to do something similar - at least I hope so (without too much effort...). It would be really useful for building automation.

Thanks again,
Andreas

**Comment by simonkowallik** [comment]
@afreudenreich I've looked into it and pushed an update to `edge`. The VS Code extension actually submits the declaration for verification and fetches the response form AS3 directly, at least in the provided examples.

The jsonschema python package is a different implementation, hence the error details varies.

Here is an example (declaration at https://github.com/simonkowallik/as3ninja/blob/edge/tests/testdata/cli/validate/errors.json):

```python
from as3ninja import schema
from as3ninja.exceptions import AS3ValidationError
as3s = schema.AS3Schema()

try:
    with open('tests/testdata/cli//validate/errors.json', "r") as f:
        declaration = f.read()
    as3s.validate(declaration=declaration)
except AS3ValidationError as validation_exception:
    for exc in validation_exception.context:
        print(exc)
        print('---')
```

The above code will produce:

```
'abc.test.com:80' does not match '^[A-Za-z]([0-9A-Za-z_.-]{0,188}[0-9A-Za-z_.])?$'

Failed validating 'pattern' in schema[0]['properties']['declaration']['additionalProperties']['additionalProperties']['propertyNames']:
    {'maxLength': 190,
     'pattern': '^[A-Za-z]([0-9A-Za-z_.-]{0,188}[0-9A-Za-z_.])?$'}

On instance['declaration']['SchemaValidationTenant']['ServiceName_Invalid']:
    'abc.test.com:80'
---
{'snat_pool': '/Common/snat_pool'} is not of type 'string'

Failed validating 'type' in schema[0]['properties']['declaration']['additionalProperties']['additionalProperties']['additionalProperties']['allOf'][0]['then']['allOf'][0]['allOf'][1]['allOf'][0]['properties']['snat']['else']['properties']['bigip']:
    {'description': 'Pathname of existing BIG-IP SNAT pool',
     'format': 'f5bigip',
     'title': 'BIG-IP SNAT pool',
     'type': 'string'}

On instance['declaration']['SchemaValidationTenant']['SNAT_Invalid2']['servicename']['snat']['bigip']:
    {'snat_pool': '/Common/snat_pool'}
---
{'snat_pool': '/Common/snat_pool'} is not a 'f5bigip'

Failed validating 'format' in schema[0]['properties']['declaration']['additionalProperties']['additionalProperties']['additionalProperties']['allOf'][0]['then']['allOf'][0]['allOf'][1]['allOf'][0]['properties']['snat']['else']['properties']['bigip']:
    {'description': 'Pathname of existing BIG-IP SNAT pool',
     'format': 'f5bigip',
     'title': 'BIG-IP SNAT pool',
     'type': 'string'}

On instance['declaration']['SchemaValidationTenant']['SNAT_Invalid2']['servicename']['snat']['bigip']:
    {'snat_pool': '/Common/snat_pool'}
---
'INVALID_SNAT_POOL_REFERENCE' is not one of ['none', 'self', 'auto']

Failed validating 'enum' in schema[0]['properties']['declaration']['additionalProperties']['additionalProperties']['additionalProperties']['allOf'][0]['then']['allOf'][0]['allOf'][1]['allOf'][0]['properties']['snat']['then']:
    {'enum': ['none', 'self', 'auto'], 'type': 'string'}

On instance['declaration']['SchemaValidationTenant']['SNAT_Invalid']['servicename']['snat']:
    'INVALID_SNAT_POOL_REFERENCE'
---
'ADC' was expected

Failed validating 'const' in schema[1]['properties']['class']:
    {'const': 'ADC',
     'description': 'Indicates this JSON document is an ADC declaration',
     'title': 'Class',
     'type': 'string'}

On instance['class']:
    'AS3'
---

...
```

I do not understand why jsonschema complains about `ADC` vs. `AS3` in the "root" `class` as well as the following "failures". This just pops up along real validation failures and is quiet otherwise.

But this does present a level of detail helpful to identify the actual issue, at least as long as a declaration is not "far too invalid".

@afreudenreich what do you think?

**Comment by afreudenreich** [comment]
@simonkowallik Thanks Simon, that looks great.

I realized one can add a line providing the schema in a declaration and VS Code/F5 extension will validate it (without submitting the declaration). This makes it easier to compare to your python code and gives pretty much the same results for your error file. 
I also don't understand the additional failures for ADC, schemaVersion etc. Might be that when a failure exists other properties also become failures (as elements are not ordered within a dict?)!? Probably hard to get rid of or filter before they end up in the context.

But as long as the other errors are shown properly that is a huge improvement and definitely usable for pipeline validation.

I have tested the edge branch code and it works as described for me. So whenever you feel comfortable with it feel free to merge into main.

Thanks again for your work - really appreciated :)

**Comment by afreudenreich** [comment]
One more update:
If I just add a break e.g. as follows:

```python
for exc in validation_exception.context:
    if exc.instance == 'AS3' and exc.json_path == '$.class' and exc.message == "'ADC' was expected":
        break
```
All the "additional" failures are gone. I have tested this with several error combinations and all the "real" errors still show. Maybe the parser does more than one round to verify the schema? But then I don't know enough about the internals of the validation...
For me this is a good enough workaround to get rid of these additional errors. I can just remove it if I notice later that this workaround hides some "real" errors.
